aggregators/codec.go

```diff
@@ -205,60 +205,67 @@ func (m *ServiceInstanceMetrics) ToProto() *aggregationpb.ServiceInstanceMetrics
 	if len(m.TransactionGroups) > cap(pb.TransactionMetrics) {
 		pb.TransactionMetrics = make([]*aggregationpb.KeyedTransactionMetrics, 0, len(m.TransactionGroups))
 	}
-	for k, m := range m.TransactionGroups {
-		ktm := aggregationpb.KeyedTransactionMetricsFromVTPool()
-		ktm.Key = k.ToProto()
-		ktm.Metrics = m.ToProto()
-		pb.TransactionMetrics = append(pb.TransactionMetrics, ktm)
+	for _, m := range m.TransactionGroups {
+		pb.TransactionMetrics = append(pb.TransactionMetrics, m)
 	}
 	if len(m.ServiceTransactionGroups) > cap(pb.ServiceTransactionMetrics) {
 		pb.ServiceTransactionMetrics = make([]*aggregationpb.KeyedServiceTransactionMetrics, 0, len(m.ServiceTransactionGroups))
 	}
-	for k, m := range m.ServiceTransactionGroups {
-		kstm := aggregationpb.KeyedServiceTransactionMetricsFromVTPool()
-		kstm.Key = k.ToProto()
-		kstm.Metrics = m.ToProto()
-		pb.ServiceTransactionMetrics = append(pb.ServiceTransactionMetrics, kstm)
+	for _, m := range m.ServiceTransactionGroups {
+		pb.ServiceTransactionMetrics = append(pb.ServiceTransactionMetrics, m)
 	}
 	if len(m.SpanGroups) > cap(pb.SpanMetrics) {
 		pb.SpanMetrics = make([]*aggregationpb.KeyedSpanMetrics, 0, len(m.SpanGroups))
 	}
-	for k, m := range m.SpanGroups {
-		ksm := aggregationpb.KeyedSpanMetricsFromVTPool()
-		ksm.Key = k.ToProto()
-		ksm.Metrics = m.ToProto()
-		pb.SpanMetrics = append(pb.SpanMetrics, ksm)
+	for _, m := range m.SpanGroups {
+		pb.SpanMetrics = append(pb.SpanMetrics, m)
 	}
 	return pb
 }
 
 // FromProto converts protobuf representation to ServiceInstanceMetrics.
 func (m *ServiceInstanceMetrics) FromProto(pb *aggregationpb.ServiceInstanceMetrics) {
-	m.TransactionGroups = make(map[TransactionAggregationKey]TransactionMetrics, len(pb.TransactionMetrics))
-	for _, ktm := range pb.TransactionMetrics {
+	m.TransactionGroups = make(
+		map[TransactionAggregationKey]*aggregationpb.KeyedTransactionMetrics,
+		len(pb.TransactionMetrics),
+	)
+	for i := range pb.TransactionMetrics {
+		ktm := pb.TransactionMetrics[i]
 		var k TransactionAggregationKey
-		var v TransactionMetrics
 		k.FromProto(ktm.Key)
-		v.FromProto(ktm.Metrics)
-		m.TransactionGroups[k] = v
-	}
-	m.ServiceTransactionGroups = make(map[ServiceTransactionAggregationKey]ServiceTransactionMetrics,
-		len(pb.ServiceTransactionMetrics))
-	for _, kstm := range pb.ServiceTransactionMetrics {
+		m.TransactionGroups[k] = ktm
+		// TODO: Either clone proto or add a comment that we change the input
```

Comment:
Even better would be to get rid of the FromProto methods. I think we should be able to do that if we:
 - create an empty CombinedMetrics when constructing a new merger, and merge the initial metrics into that (rather than using FromProto for the first metrics)
 - change the Processor function type to accept an `*aggregationpb.CombinedMetrics`

I think ideally we would make CombinedMetrics and friends unexported types, and only use them for merging. Perhaps we should move all merging logic to an internal package. WDYT?

The second point might be a heavy lift. If it is, then I think as an interim step we could still remove FromProto by merging onto an empty CombinedMetrics with no limits.

Reply:
BTW that doesn't need to happen in this PR. I can create a followup if you think it sounds reasonable.

Reply:
That sounds reasonable, I will give it a go for point 1. Point 2 also I think is simple, will create a followup for that.

Reply:
Also, love the idea of hinding the full model (minus the `*Key` structs). Will give that a go to.

BTW, `*Key` need to stay since the proto models are not comparable. This is because the proto models have some private fields ([ref](https://github.com/golang/protobuf/issues/531)).

Reply:
Discussed with @axw offline, decided to create follow up PRs.

Reply:
> change the Processor function type to accept an *aggregationpb.CombinedMetrics

I had the same idea. Agg key map access aesbodyhash is taking some time now, but can't be sure if performance is better without maps (sorted array merge) or with a temporary map (most likely worse).

> BTW, *Key need to stay since the proto models are not comparable. This is because the proto models have some private fields

Could `EqualVT` be useful?